File: apps/android/app/src/main/kotlin/sh/nikhil/swekitty/ui/InSessionBottomBar.kt
package sh.nikhil.swekitty.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AddCircle
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.outlined.Layers
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay

/**
 * Active-tab context for the in-session bottom bar — mirrors
 * `apps/ios/Sources/Views/InSessionBottomBar.swift` `InSessionContext`.
 * Drives where the centre mic FAB routes its voice transcript: chat
 * gets the existing voice path, terminal / browser surface a "not
 * supported" inline note. Lives separately from [ProjectTab] so the
 * bar's pure-data model can be unit-tested without standing up the
 * Compose tree.
 */
enum class InSessionContext {
    Terminal, Chat, Browser;

    companion object {
        /** Bridge from the segmented-tab enum so a future ProjectTab rename
         *  is caught at compile time via the exhaustive `when`. */
        fun fromTab(tab: ProjectTab): InSessionContext = when (tab) {
            ProjectTab.Terminal -> Terminal
            ProjectTab.Chat     -> Chat
            ProjectTab.Browser  -> Browser
        }
    }
}

/**
 * Pure-data description of the in-session bottom bar. Three controls
 * in a fixed leading → centre → trailing order: thread switcher
 * (Outlined.Layers, iOS `square.stack` equivalent), voice FAB
 * (Filled.Mic), new-session (Filled.AddCircle). Lifted out of the
 * composable so the unit tests in `InSessionBottomBarModelTest` can
 * pin the three-control structure + per-tab voice routing without a
 * Compose host. Same pattern as iOS `InSessionBottomBarModel`.
 */
object InSessionBottomBarModel {

    enum class Control {
        Threads,
        Voice,
        NewSession;

        /** Material icon used in the rendered bar. Asserted by tests. */
        val icon: ImageVector
            get() = when (this) {
                Threads    -> Icons.Outlined.Layers
                Voice      -> Icons.Filled.Mic
                NewSession -> Icons.Filled.AddCircle
            }

        /** Accessibility / content description label. Asserted by tests. */
        val accessibilityLabel: String
            get() = when (this) {
                Threads    -> "Switch thread"
                Voice      -> "Voice dictation"
                NewSession -> "New session"
            }
    }

    /** Render order: leading → centre → trailing. Tests pin this triple. */
    val controls: List<Control> = listOf(Control.Threads, Control.Voice, Control.NewSession)

    /**
     * Whether the centre mic FAB is wired to the existing voice path
     * for the supplied tab context. Per the spec: v1 supports chat
     * only; terminal / browser surface an inline note. Tests assert
     * the routing table so a future refactor can't silently broaden
     * or shrink the supported set.
     */
    fun voiceSupported(context: InSessionContext): Boolean = when (context) {
        InSessionContext.Chat                              -> true
        InSessionContext.Terminal, InSessionContext.Browser -> false
    }

    /** Message used by the inline note when voice isn't wired for the current tab. */
    fun voiceUnsupportedMessage(context: InSessionContext): String = "Voice not supported here"
}

/**
 * Persistent in-session bottom dock — visible across the Terminal /
 * Chat / Browser tabs so the user always has the same three controls
 * at thumb-reach: switch parallel sessions on this server (left),
 * fire voice into the active tab (centre), or spin up another session
 * on the same server (right). Mirrors iOS `InSessionBottomBar`.
 */
@Composable
fun InSessionBottomBar(
    context: InSessionContext,
    onThreads: () -> Unit,
    onVoice: () -> Unit,
    onNewSession: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var unsupportedNote by remember { mutableStateOf<String?>(null) }

    // Auto-clear the unsupported toast so the bar returns to its
    // three-icon resting state — same 1.8s window as iOS.
    LaunchedEffect(unsupportedNote) {
        if (unsupportedNote != null) {
            delay(1_800)
            unsupportedNote = null
        }
    }

    Column(
        modifier = modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        AnimatedVisibility(visible = unsupportedNote != null, enter = fadeIn(), exit = fadeOut()) {
            Text(
                text = unsupportedNote.orEmpty(),
                style = MaterialTheme.typography.labelMedium,
                color = SweKittyTheme.textPrimary(),
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .padding(bottom = 6.dp)
                    .glassCapsule(tint = SweKittyTheme.warning().copy(alpha = 0.34f))
                    .padding(horizontal = 12.dp, vertical = 6.dp)
                    .semantics { contentDescription = "InSessionBottomBar.voiceUnsupported" },
            )
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 14.dp, vertical = 4.dp)
                .glassCapsule(tint = SweKittyTheme.surface().copy(alpha = 0.55f))
                .padding(horizontal = 22.dp, vertical = 10.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            DockCircleButton(
                icon = InSessionBottomBarModel.Control.Threads.icon,
                label = InSessionBottomBarModel.Control.Threads.accessibilityLabel,
                tint = SweKittyTheme.textPrimary(),
                onClick = onThreads,
            )
            Spacer(Modifier.weight(1f))
            VoiceFab(
                onClick = {
                    if (InSessionBottomBarModel.voiceSupported(context)) {
                        onVoice()
                    } else {
                        unsupportedNote = InSessionBottomBarModel.voiceUnsupportedMessage(context)
                    }
                },
            )
            Spacer(Modifier.weight(1f))
            DockCircleButton(
                icon = InSessionBottomBarModel.Control.NewSession.icon,
                label = InSessionBottomBarModel.Control.NewSession.accessibilityLabel,
                tint = SweKittyTheme.accentStrong(),
                onClick = onNewSession,
            )
        }
    }
}

@Composable
private fun DockCircleButton(
    icon: ImageVector,
    label: String,
    tint: androidx.compose.ui.graphics.Color,
    onClick: () -> Unit,
) {
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(44.dp)
            .clip(CircleShape)
            .glassCircle(tint = SweKittyTheme.surface().copy(alpha = 0.7f))
            .clickable(onClick = onClick)
            .semantics { contentDescription = label },
    ) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(22.dp))
    }
}

/**
 * Centre mic FAB — bigger and copper-accented to match iOS's dominant
 * voice affordance. Disabled state is handled by the caller via
 * [InSessionBottomBarModel.voiceSupported] so the button itself only
 * fires `onClick`.
 */
@Composable
private fun VoiceFab(onClick: () -> Unit) {
    val accent = SweKittyTheme.accentStrong()
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(60.dp)
            .clip(CircleShape)
            .background(accent, CircleShape)
            .border(width = 3.dp, color = accent.copy(alpha = 0.55f), shape = CircleShape)
            .clickable(onClick = onClick)
            .semantics { contentDescription = InSessionBottomBarModel.Control.Voice.accessibilityLabel },
    ) {
        Icon(
            InSessionBottomBarModel.Control.Voice.icon,
            contentDescription = null,
            tint = SweKittyTheme.textOnAccent(),
            modifier = Modifier.size(28.dp),
        )
    }
}

File: apps/ios/Sources/Views/InSessionBottomBar.swift
import SwiftUI

/// Active-tab context for the in-session bottom bar. Drives where the
/// voice transcript is routed when the user taps the center mic FAB —
/// chat gets `sendChat`, terminal gets `sendInput` (line-terminated),
/// browser falls through to a "not supported" toast for v1. Mirrors
/// `ProjectTab` but lives separately so the bar can be unit-tested
/// without standing up the SwiftUI view tree.
enum InSessionContext: String, CaseIterable, Equatable {
    case terminal
    case chat
    case browser

    init(_ tab: ProjectTab) {
        switch tab {
        case .terminal: self = .terminal
        case .chat:     self = .chat
        case .browser:  self = .browser
        }
    }
}

/// Pure-data description of the in-session bottom bar. Three controls
/// in a fixed left → center → trailing order — `square.stack` (thread
/// switcher), `mic.fill` (voice FAB), `plus.circle.fill` (new session
/// on the same server). Lifted out of the SwiftUI view so the tests
/// in `InSessionBottomBarTests` can pin the three-control structure
/// without a host controller. Same pattern as PR B's
/// `ProjectHeaderModel`.
struct InSessionBottomBarModel: Equatable {
    enum Control: String, CaseIterable, Equatable {
        case threads
        case voice
        case newSession

        /// SF Symbol used in the rendered bar. Asserted by tests so a
        /// refactor can't quietly swap the affordance.
        var systemImage: String {
            switch self {
            case .threads:    return "square.stack"
            case .voice:      return "mic.fill"
            case .newSession: return "plus.circle.fill"
            }
        }

        /// Accessibility label for VoiceOver. Asserted by tests.
        var accessibilityLabel: String {
            switch self {
            case .threads:    return "Switch thread"
            case .voice:      return "Voice dictation"
            case .newSession: return "New session"
            }
        }
    }

    /// Render order: leading → center → trailing. Tests assert this
    /// exact triple so the litter HomeBottomBar visual parity holds.
    static let controls: [Control] = [.threads, .voice, .newSession]

    /// Whether the center mic FAB is wired to the existing voice path
    /// for the supplied tab context. Per the spec: v1 supports chat
    /// only; terminal / browser surface a toast instead. Tests assert
    /// the routing table so a future refactor can't silently broaden
    /// or shrink the supported set.
    static func voiceSupported(for context: InSessionContext) -> Bool {
        switch context {
        case .chat:                  return true
        case .terminal, .browser:    return false
        }
    }

    /// Message used by the toast / inline note when voice isn't wired
    /// for the current tab. Mono-cased for parity with our other UI
    /// strings.
    static func voiceUnsupportedMessage(for context: InSessionContext) -> String {
        "Voice not supported here"
    }
}

/// Persistent in-session bottom dock — visible across the Terminal /
/// Chat / Browser tabs so the user always has the same three controls
/// at thumb-reach: switch parallel sessions on this server (left),
/// fire voice dictation into the active tab (center), or spin up
/// another session on the same server (right). Mirrors litter's
/// `HomeBottomBar`, but scoped to within a session — `BottomActionBar`
/// already covers the home surface.
struct InSessionBottomBar: View {
    let context: InSessionContext
    let onThreads: () -> Void
    let onVoice: () -> Void
    let onNewSession: () -> Void

    /// One-shot transient message surfaced when the user taps voice
    /// on an unsupported tab. Cleared after a short delay so the bar
    /// goes back to its three-icon resting state.
    @State private var unsupportedNote: String?

    var body: some View {
        VStack(spacing: 6) {
            if let note = unsupportedNote {
                Text(note)
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(SweKittyTheme.textPrimary)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 6)
                    .glassCapsule(tint: SweKittyTheme.warning.opacity(0.34))
                    .transition(.opacity)
                    .accessibilityIdentifier("InSessionBottomBar.voiceUnsupported")
            }

            HStack(spacing: 14) {
                threadsButton
                Spacer()
                voiceFab
                Spacer()
                newSessionButton
            }
            .padding(.horizontal, 22)
            .padding(.vertical, 10)
            .frame(maxWidth: .infinity)
            // Glass capsule mirrors litter's HomeBottomBar dock shape —
            // the row is wide enough that the capsule's pill ends read
            // as soft shoulders, not aggressive ovals.
            .glassCapsule(tint: SweKittyTheme.surface.opacity(0.55))
            .padding(.horizontal, 14)
            .padding(.bottom, 6)
        }
    }

    private var threadsButton: some View {
        Button(action: onThreads) {
            Image(systemName: InSessionBottomBarModel.Control.threads.systemImage)
                .font(.title3.weight(.semibold))
                .foregroundStyle(SweKittyTheme.textPrimary)
                .frame(width: 44, height: 44)
                .glassCircle(tint: SweKittyTheme.surface.opacity(0.7))
        }
        .buttonStyle(.plain)
        .accessibilityLabel(InSessionBottomBarModel.Control.threads.accessibilityLabel)
    }

    /// Center mic FAB — bigger and copper-accented to match litter's
    /// dominant voice affordance. Tap routes to `onVoice` when the
    /// tab supports voice, otherwise surfaces the "not supported" note
    /// so the user gets feedback instead of a no-op.
    private var voiceFab: some View {
        Button {
            if InSessionBottomBarModel.voiceSupported(for: context) {
                onVoice()
            } else {
                let msg = InSessionBottomBarModel.voiceUnsupportedMessage(for: context)
                withAnimation(.easeInOut(duration: 0.15)) {
                    unsupportedNote = msg
                }
                Task { @MainActor in
                    try? await Task.sleep(nanoseconds: 1_800_000_000)
                    withAnimation(.easeInOut(duration: 0.2)) {
                        unsupportedNote = nil
                    }
                }
            }
        } label: {
            Image(systemName: InSessionBottomBarModel.Control.voice.systemImage)
                .font(.title.weight(.bold))
                .foregroundStyle(SweKittyTheme.textOnAccent)
                .frame(width: 60, height: 60)
                .background(
                    Circle()
                        .fill(SweKittyTheme.accentStrong)
                )
                .overlay(
                    Circle()
                        .stroke(SweKittyTheme.accentStrong.opacity(0.55), lineWidth: 3)
                        .blur(radius: 2)
                )
        }
        .buttonStyle(.plain)
        .accessibilityLabel(InSessionBottomBarModel.Control.voice.accessibilityLabel)
    }

    private var newSessionButton: some View {
        Button(action: onNewSession) {
            Image(systemName: InSessionBottomBarModel.Control.newSession.systemImage)
                .font(.title2.weight(.semibold))
                .foregroundStyle(SweKittyTheme.accentStrong)
                .frame(width: 44, height: 44)
                .glassCircle(tint: SweKittyTheme.surface.opacity(0.7))
        }
        .buttonStyle(.plain)
        .accessibilityLabel(InSessionBottomBarModel.Control.newSession.accessibilityLabel)
    }
}
